### PR TITLE
Add unsubmit feature and status field in grading overview

### PR DESCRIPTION
**Pull request by geshuming**
- Added unsubmit feature
Avengers can only unsubmit submissions made by students in their group, hence the unsubmit button becomes hidden when fetching all submissions.

- Added status on the grading overview.
Avengers should be able to tell if the students are still attempting/submitted their assessment.

### CODE REVIEW

**Comment by geshuming** [comment]
Ok, I reverted the style changes and separated them into a new PR.

**Comment by geshuming** [comment]
@rrtheonlyone jiachen recommended to hide the unsubmit button when the grader chooses to see all submissions. Currently, the feature in the backend only allows avengers of a student's group to unsubmit, even profs cannot unsubmit. People without a class will also be affected.

**Comment by rrtheonlyone** [comment]
Interesting. Why not do it like this?

- If role is `:admin`, he/she should be able to unsubmit regardless (`:admin` is for profs/senior teaching staff)
- If role is `:staff`, he/she can only unsubmit for their own class

This is just a contingency just in case the TA happens to not be around or if the prof needs to be step in. @martin-henz @jiachen247 @geshuming let me know your thoughts on this!

**Comment by martin-henz** [comment]
> Interesting. Why not do it like this?
> 
> * If role is `:admin`, he/she should be able to unsubmit regardless (`:admin` is for profs/senior teaching staff)
> * If role is `:staff`, he/she can only unsubmit for their own class
> 
> This is just a contingency just in case the TA happens to not be around or if the prof needs to be step in. @martin-henz @jiachen247 @geshuming let me know your thoughts on this!

Yes, that sounds great.

**Comment by rrtheonlyone** [comment]
Okay @geshuming / @jiachen247  could you handle the updating of the backend logic for this? 

With this implemented, we do not need to hide the `unsubmit` button on the frontend side when viewing all submissions.

**Comment by geshuming** [comment]
@rrtheonlyone ive reverted the hiding of the unsubmit column. there should also be a PR to allow admin to unsubmit

**Comment by rrtheonlyone** [comment]
Okay sounds good 👍 

Will take a look soon

File: src/components/academy/grading/index.tsx [hunk 1]
```diff
@@ -25,6 +34,7 @@ type State = {
   columnDefs: ColDef[];
   filterValue: string;
   groupFilterEnabled: boolean;
+  unsubmitAlert: JSX.Element | null;
```
Comment (rrtheonlyone) [inline src/components/academy/grading/index.tsx]: Why is the `unsubmitAlert` part of the state?

File: src/components/academy/grading/index.tsx [hunk 2]
```diff
@@ -242,6 +284,34 @@ class Grading extends React.Component<IGradingProps, State> {
     }
     this.gridApi.exportDataAsCsv({ allColumns: true });
   };
+
+  private handleUnsubmitAlertClose = () => this.setState({ unsubmitAlert: null });
+
+  private unsubmitButton = (props: GradingNavLinkProps) => {
```
Comment (rrtheonlyone) [inline src/components/academy/grading/index.tsx]: I feel this state change is rather unnecessary. It is adding unnecessary verbosity to your code and the `null` change feels hacky. Possible to look into a cleaner alternative?

File: src/components/academy/grading/index.tsx [hunk 3]
```diff
@@ -242,6 +284,34 @@ class Grading extends React.Component<IGradingProps, State> {
...
+        onConfirm={handleConfirmUnsubmit}
+        isOpen={true}
+        onCancel={this.handleUnsubmitAlertClose}
+      >
+        <p>Are you sure you want to unsubmit this submission?</p>
+      </Alert>
+    );
+
+    return controlButton('', IconNames.ARROW_LEFT, () => this.setState({ unsubmitAlert }), {
+      fullWidth: true
```
Comment (rrtheonlyone) [inline src/components/academy/grading/index.tsx]: What is this doing?

File: src/mocks/backend.ts [hunk 4]
```diff
@@ -82,6 +82,11 @@ export function* mockBackendSaga(): SagaIterator {
     yield call(showSuccessMessage, 'Saved!', 1000);
   });
 
+  yield takeEvery(actionTypes.UNSUBMIT_SUBMISSION, function*(action) {
+    /* TODO */
```
Comment (rrtheonlyone) [inline src/mocks/backend.ts]: Are you planning on adding this in later?

File: src/sagas/backend.ts [hunk 5]
```diff
@@ -199,6 +199,43 @@ function* backendSaga(): SagaIterator {
...
+    const tokens = yield select((state: IState) => ({
+      accessToken: state.session.accessToken,
+      refreshToken: state.session.refreshToken
+    }));
+    const { submissionId, overviews } = (action as actionTypes.IAction).payload;
+    const resp: Response = yield postUnsubmit(submissionId, tokens);
+    const newOverviews = (overviews as GradingOverview[]).map(overview => {
+      if (overview.submissionId === submissionId) {
+        return { ...overview, submissionStatus: 'attempted' };
+      } else {
```
Comment (rrtheonlyone) [inline src/sagas/backend.ts]: No need for the `else` here

File: src/sagas/backend.ts [hunk 6]
```diff
@@ -199,6 +199,43 @@ function* backendSaga(): SagaIterator {
...
+        return overview;
+      }
+    });
+
+    if (resp && resp.ok) {
+      yield call(showSuccessMessage, 'Unsubmit successful', 1000);
+      yield put(actions.updateGradingOverviews(newOverviews));
+    } else if (resp !== null) {
+      let errorMessage: string;
```
Comment (rrtheonlyone) [inline src/sagas/backend.ts]: This block is duplicated many times in the code. Let's maintain DRY principle: Please abstract it out into a helper function.

Comment (geshuming) [reply src/sagas/backend.ts]: should we handle it now or log an issue. I can see that the entire file benefits from a `handleError` function, but I don't this PR will be suitable to change that

File: src/components/academy/grading/index.tsx (hunk 3 above)
Comment (geshuming) [reply src/components/academy/grading/index.tsx]: `fullWidth` is an option for the `controlButton` component, setting the button's width to 100% i believe

File: src/mocks/backend.ts (hunk 4 above)
Comment (geshuming) [reply src/mocks/backend.ts]: Yes, done

File: src/sagas/backend.ts (hunk 5 above)
Comment (geshuming) [reply src/sagas/backend.ts]: Removed

File: src/components/academy/grading/index.tsx (hunk 2 above)
Comment (geshuming) [reply src/components/academy/grading/index.tsx]: Removed, isolated the unsubmit cell component

File: src/components/academy/grading/index.tsx (hunk 1 above)
Comment (geshuming) [reply src/components/academy/grading/index.tsx]: Removed, isolated the unsubmit cell component

File: src/sagas/backend.ts (hunk 6 above)
Comment (rrtheonlyone) [reply src/sagas/backend.ts]: We should handle it now? I don't see it as a big change to warrant a new PR. (its just refactoring?)

Your PR is actually adding to the issue and making it worse! So it is something that needs to be fixed. May I know why you would rather log it as an issue?

Comment (geshuming) [reply src/sagas/backend.ts]: Ok, I abstracted it out

File: src/sagas/backend.ts [hunk 7]
```diff
@@ -201,6 +201,32 @@ function* backendSaga(): SagaIterator {
...
+   * Unsubmits the submission and updates the grading overviews of the new status.
+   */
+  yield takeEvery(actionTypes.UNSUBMIT_SUBMISSION, function*(action) {
+    const tokens = yield select((state: IState) => ({
+      accessToken: state.session.accessToken,
+      refreshToken: state.session.refreshToken
+    }));
+    const { submissionId } = (action as actionTypes.IAction).payload;
+    const overviews = yield select((state: IState) => state.session.gradingOverviews || []);
+    const resp: Response = yield postUnsubmit(submissionId, tokens);
```
Comment (rrtheonlyone) [inline src/sagas/backend.ts]: Again use a guard clause here. It is more efficient. If not, you are unnecessary looping through the `overviews` when your resp has returned an error.

```
if (!resp || !resp.ok) {
     //handle error
}

//main logic
```

File: src/sagas/backend.ts [hunk 8]
```diff
@@ -600,6 +629,23 @@ async function request(
   }
 }
 
+function* handleResponseError(resp: Response | null) {
+  if (resp) {
```
Comment (rrtheonlyone) [inline src/sagas/backend.ts]: Use a guard clause similar to above^

File: src/components/academy/grading/index.tsx [hunk 9]
```diff
@@ -25,6 +27,7 @@ type State = {
   columnDefs: ColDef[];
   filterValue: string;
   groupFilterEnabled: boolean;
+  unsubmitAlert: JSX.Element | null;
```
[the remark this answers is not in the document]
Comment (geshuming) [reply src/components/academy/grading/index.tsx]: All removed

File: src/mocks/backend.ts [hunk 10]
```diff
@@ -82,6 +86,29 @@ export function* mockBackendSaga(): SagaIterator {
...
+  yield takeEvery(actionTypes.UNSUBMIT_SUBMISSION, function*(action) {
+    const { submissionId } = (action as actionTypes.IAction).payload;
+    const overviews: GradingOverview[] = yield select(
+      (state: IState) => state.session.gradingOverviews || []
+    );
+    const index = overviews.findIndex(
+      overview =>
+        overview.submissionId === submissionId && overview.submissionStatus === 'submitted'
+    );
+    if (index >= 0) {
```
[the remark this answers is not in the document]
Comment (geshuming) [reply src/mocks/backend.ts]: Done

File: src/sagas/backend.ts (hunk 7 above)
Comment (geshuming) [reply src/sagas/backend.ts]: Done

File: src/mocks/backend.ts [hunk 11]
```diff
@@ -82,6 +86,29 @@ export function* mockBackendSaga(): SagaIterator {
...
+      const newOverviews = (overviews as GradingOverview[]).map(overview => {
+        if (overview.submissionId === submissionId) {
+          return { ...overview, submissionStatus: 'attempted' };
+        }
+        return overview;
+      });
+      yield call(showSuccessMessage, 'Unsubmitted!', 1000);
+      yield put(actions.updateGradingOverviews(newOverviews));
+    } else {
+      yield call(showWarningMessage, 'Cannot unsubmit! Submission not submitted!');
```
[the remark this answers is not in the document]
Comment (geshuming) [reply src/mocks/backend.ts]: Nope. I changed it to "400: Bad Request" to align with backend

File: src/components/academy/grading/index.tsx [hunk 12]
```diff
@@ -108,12 +112,44 @@ class Grading extends React.Component<IGradingProps, State> {
...
         },
+        {
+          headerName: 'Status',
+          field: 'submissionStatus',
+          valueFormatter: (params: ValueFormatterParams) => {
+            const str = params.value as string;
+            return str.charAt(0).toUpperCase() + str.slice(1);
+          },
+          cellStyle: (params: GradingNavLinkProps) => {
+            if (params.data.submissionStatus === 'submitted') {
```
[the remark this answers is not in the document]
Comment (geshuming) [reply src/components/academy/grading/index.tsx]: Done

File: src/sagas/backend.ts [hunk 13]
```diff
@@ -600,6 +630,24 @@ async function request(
...
+    return;
+  }
+
+  let errorMessage: string;
+  switch (resp.status) {
+    case 401:
+      errorMessage = 'Session expired. Please login again.';
+      break;
+    default:
+      errorMessage = `Error ${resp.status}: ${resp.statusText}`;
```
Comment (rrtheonlyone) [inline src/sagas/backend.ts]: Might want to consider removing the `status` since it does not add any value to the end user.

Comment (geshuming) [reply src/sagas/backend.ts]: I think it should be fine, certainly can use better error messages

**Pull request merged**